packages/core/src/types/errors.ts

```diff
@@ -356,6 +357,12 @@ export interface CanvasNonexistentDimsError {
   type?: Expr<A>["tag"];
 }
 
+export interface CyclicAssignmentError {
+  tag: "CyclicAssignmentError";
+  // TODO: improve types, currently the generated id and source location
```

Comment:
Probably remove this, unless you had a specific improvement in mind?

```suggestion
```

Reply:
I was hoping to make `cycles: Expr<C>[][]`, but I guess there might be some exceptions to this?

Reply:
Ah gotcha; yeah we can't actually do that, because some `DepGraph` nodes represent shapes rather than shape properties, and in general there isn't a single expression representing a shape, because `override` allows people to replace a single property of a shape while leaving the rest of the shape unchanged:

https://github.com/penrose/penrose/blob/654571a52fd347c4078db95da277e7c231bacbf6/packages/core/src/types/styleSemantics.ts#L149-L154